### ISSUE TITLE
Infinite RecursionError with inherited models

### ISSUE THREAD

**Issue by BrendaH** [rparent/django-lock-tokens#10]
* django-lock-tokens version: 0.2.4
* Django version: 1.11.16
* Python version: 3.5
* Operating System: Windows 10

### Description
models.py:
```
class ParentModel(LockableModel):
    # some fields

class ChildModel(ParentModel):
    # some extra fields
```

admin.py:
```
class ParentModelAdmin(LockableModelAdmin):
    list_display = ['function_link_name']
    list_display_links = None
    # some other stuff

    def function_link_name(self, obj):
        if obj.is_locked():
            return mark_safe(_('Someone is already editing %(name)s') % {'name': str(obj)})
        else:
            change_url = reverse('admin:xxxx_change', args=(obj.id,))  
            return mark_safe('<a class="grp-button" href="' + change_url + '" target="blank">' + str(obj) + '</a>')


class ChildModelAdmin(ParentModelAdmin):
   # some other stuff here
```

When trying to load the changelist of ChildModel in the admin I get a Recursion Error:
"RecursionError: maximum recursion depth exceeded while calling a Python object"

ChildModel seems to be calling ParentModel, which in turn calls ChildModel etc.

### Possible fix
My partner (@matthijskooijman) helped me kind of understand this issue and has a sort-of-working solution.
The offending line was `obj = getattr(instance, key)`. Which for ChildModel looks up the attribute `parentmodel_ptr`, which resolves to a ParentModel object which gets a call to getattr which follows the reverse link to the child and in turn gets a ChildModel object to call getattr on etc.

In `utils.py`, changes in `class_or_bound_method`, `DLTModelProxyBase.__call__` and `DLTModelProxyBase._bind_methods`
```
def class_or_bound_method(function):
    cm = classmethod(function)
    cm._bind_at_instanciation = True
    cm._orig_f = function
    return cm

class DLTModelProxyBase(models.base.ModelBase):

    def __call__(cls, *args, **kwargs):
        instance = super(DLTModelProxyBase, cls).__call__(*args, **kwargs)
        cls._bind_methods(instance, instance.__class__)
        return instance

    @classmethod
    def _bind_methods(cls, instance, instance_class):
        for (key, value) in instance_class.__dict__.items():
            if getattr(value, '_bind_at_instanciation', False):
                obj = getattr(instance, key)
                setattr(instance, key, cls._get_bound_method(obj, instance))
        for base in instance_class.__bases__:
            cls._bind_methods(instance, base)
```
(Sidenote: instantiation is spelled wrong with a 'c'.)

This solution is not perfect and @matthijskooijman wants to take a better look at it later. For now this lets me continue the work I was doing because the error is gone, but without a fix and new release of django-lock-tokens I can not put it on the production server so I am looking forward to a neat solution. (Unfortunately my understanding of Python is not sufficient to help with this :-s )

**Comment by matthijskooijman** [comment]
As promised, a bit more info from me (though I won't have time to improve the solution, but will comment on it).

The problem here is that `DLTModelProxyBase` iterates over all properties of a model using getattr. Normally, that seems harmless, but in case of dynamic properties ([descriptors](https://docs.python.org/3/howto/descriptor.html)), this can actually trigger code to run. In this case, there is a problem with an infinite recursion through the parent ptr property, but in general it seems a bad idea to force evaluating of attributes in this way (it forces loading of on-demand cached properties, for example).

In this particular case, the problem is (IIRC) that the parent ptr attribute returns an instance of the parent class, which *also* gets its properties evaluated, including the reverse property of the parent ptr, which returns a (new instance of the) child class, etc.

The solution I had in mind was to not call `getattr` on all attributes, but rather get the raw attribute value (skipping the call to `__get__` that `getattr` when a value turns out to be a descriptor) and then look at that value. However, the `classname` decorator also returns a descriptor, so I had to rewrite the code a bit to put the `_bind_at_instanciation` attribute on the descriptor, rather than the original function. Since the class method lives in the class object (rather than the instance), `_bind_methods` iterates up into the class hierarchy to find the class methods and puts a bound version on them on the instance.

However, this has a few problems:
 - Essentially, the suggested code duplicates the resolution logic that `getattr` does (looking through the chain of superclasses). However, it does not really handle attributes that are overridden in subclasses (it might end up binding a superclass method, ignoring the override in a subclass). Also, I think multiple inheritance are not really handled properly.
 - The code still runs `getattr` to get at the original function. I tried using the original function, but that would require changes to `_get_bound_method` as well, so this was just the easiest approach.

Going in entirely a different direction, perhaps the metaclass could be removed and the entire thing can be handled in a descriptor? IIUC, `class_or_bound_method` could set a descriptor object (essentially re-implementing the classmethod decorator). The `__get__` method callaed on this descriptor object gets called with an instance and class object, where the instance argument is `None` when the method is called as a class method.

See https://docs.python.org/3/howto/descriptor.html

**Comment by BrendaH** [comment]
Is there any progress on this issue? I am nearing the day that the branch in my project has to be merged and everything put on production, but that will break without a fix for this issue.

**Comment by rparent** [comment]
Hello @BrendaH, and sorry for the late answer.

Unfortunately I have been very busy lately and did not have any time to look at this issue in detail, as it is not trivial. I think using descriptors as suggested by @matthijskooijman is the good way to go. I will try to have a look at it as soon as possible.

In the meantime, feel free to submit a PR, if you really are in a hurry. I will review and merge it much faster.

**Comment by rparent** [comment]
Hello @BrendaH 

I just commited a fix to your problem, can you test it to see if it works ? If it does fix the problem, I will release a new version of the package.

**Comment by matthijskooijman** [comment]
Nice! I had a quick look at the code, which looks good. I did notice a lot of completely unrelated formatting changes to the code (and one actual change wrt six / python2 compatibility) in the same commit, that made reviewing this a bit more tricky, but I think I managed :-)

**Comment by rparent** [comment]
Ok perfect, I just released `v0.2.5` on PyPI!
